File: typescript-pixi/MapBuilder.ts
/// <reference path="SliceType.ts" />
/// <reference path="Walls.ts" />

class MapBuilder {

	private walls :Walls;

	private static WALL_HEIGHTS :number[] = [
		256, // Lowest slice
		224,
		192,
		160,
		128  // Highest slice
	];

	constructor(walls :Walls) {
		this.walls = walls;
		this.createMap();
	}

	private createMap() :void {
		this.createWallSpan(3, 9, true);
		this.createGap(1);
		this.createWallSpan(1, 30);
		this.createGap(1);
		this.createWallSpan(2, 18);
		this.createGap(1);
		this.createSteppedWallSpan(2, 5, 28);
		this.createGap(1);
		this.createWallSpan(1, 10);
		this.createGap(1);
		this.createWallSpan(2, 6); 
		this.createGap(1);
		this.createWallSpan(1, 8);
		this.createGap(1);
		this.createWallSpan(2, 6);
		this.createGap(1);
		this.createWallSpan(1, 8);
		this.createGap(1)
		this.createWallSpan(2, 7);
		this.createGap(1);
		this.createWallSpan(1, 16);
		this.createGap(1);
		this.createWallSpan(2, 6);
		this.createGap(1);
		this.createWallSpan(1, 22);
		this.createGap(2);
		this.createWallSpan(2, 14);
		this.createGap(2);
		this.createWallSpan(3, 8);
		this.createGap(2);
		this.createSteppedWallSpan(3, 5, 12);
		this.createGap(3);
		this.createWallSpan(0, 8);
		this.createGap(3);
		this.createWallSpan(1, 50);
		this.createGap(20);
	}

	private createGap(spanLength :number) :void {
		for (var i :number = 0; i < spanLength; i++)
		{
			this.walls.addSlice(SliceType.GAP);
		}
	}

	private createWallSpan(
		heightIndex :number,
		spanLength  :number,
		noFront     :boolean = false,
		noBack      :boolean = false
	) :void
	{
		if (noFront == false && spanLength > 0)
		{
			this.addWallFront(heightIndex);
			spanLength--;
		}

		var midSpanLength :number = spanLength - (noBack ? 0 : 1);
		if (midSpanLength > 0)
		{
			this.addWallMid(heightIndex, midSpanLength)
			spanLength -= midSpanLength;
		}

		if (noBack == false && spanLength > 0)
		{
			this.addWallBack(heightIndex);
		}
	}

	private createSteppedWallSpan(
		heightIndex :number,
		spanALength :number,
		spanBLength :number
	) :void
	{
		if (heightIndex < 2)
		{
			heightIndex = 2;
		}

		this.createWallSpan(heightIndex, spanALength, false, true);
		this.addWallStep(heightIndex - 2);
		this.createWallSpan(heightIndex - 2, spanBLength - 1, true, false);
	}

	private addWallFront(heightIndex :number) :void {
		var y :number = MapBuilder.WALL_HEIGHTS[heightIndex];
		this.walls.addSlice(SliceType.FRONT, y);
	}

	private addWallBack(heightIndex :number) :void {
		var y :number = MapBuilder.WALL_HEIGHTS[heightIndex];
		this.walls.addSlice(SliceType.BACK, y);
	}

	private addWallMid(heightIndex :number, spanLength :number) :void {
		var y :number = MapBuilder.WALL_HEIGHTS[heightIndex];
		for (var i :number = 0; i < spanLength; i++)
		{
			if (i % 2 == 0)
			{
				this.walls.addSlice(SliceType.WINDOW, y);
			}
			else
			{
				this.walls.addSlice(SliceType.DECORATION, y);
			}
		}
	}

	private addWallStep(heightIndex :number) :void {
		var y :number = MapBuilder.WALL_HEIGHTS[heightIndex];
		this.walls.addSlice(SliceType.STEP, y);
	}
}

File: swift-spritekit/Endless-Runner/MapBuilder.swift
//
//  MapBuilder.swift
//  Endless-Runner
//
//  Created by Christopher Caleb on 13/06/2014.
//  Copyright (c) 2014 Christopher Caleb. All rights reserved.
//

import Foundation
import SpriteKit

class MapBuilder {
    let WALL_HEIGHTS: CGFloat[] = [
        -256, // Lowest slice
        -192,
        -128,
         -64,
           0  // Highest slice
    ]
    
    let _walls :Walls
    
    init(walls: Walls) {
        _walls = walls
        createMap()
    }
    
    func createMap() {
        createWallSpanWithHeight(3, spanLength: 9, noFront: true)
        createGap(1)
        createWallSpanWithHeight(1, spanLength: 30)
        createGap(1)
        createWallSpanWithHeight(2, spanLength: 18)
        createGap(1)
        createSteppedWallSpanWithHeight(2, spanALength: 5, spanBLength: 28)
        createGap(1)
        createWallSpanWithHeight(1, spanLength: 10)
        createGap(1)
        createWallSpanWithHeight(2, spanLength: 6)
        createGap(1)
        createWallSpanWithHeight(1, spanLength: 8)
        createGap(1)
        createWallSpanWithHeight(2, spanLength: 6)
        createGap(1)
        createWallSpanWithHeight(1, spanLength: 8)
        createGap(1)
        createWallSpanWithHeight(2, spanLength: 7)
        createGap(1)
        createWallSpanWithHeight(1, spanLength: 16)
        createGap(1)
        createWallSpanWithHeight(2, spanLength: 6)
        createGap(1)
        createWallSpanWithHeight(1, spanLength: 22)
        createGap(2)
        createWallSpanWithHeight(2, spanLength: 14)
        createGap(2)
        createWallSpanWithHeight(3, spanLength: 8)
        createGap(2)
        createSteppedWallSpanWithHeight(3, spanALength: 5, spanBLength: 12)
        createGap(3)
        createWallSpanWithHeight(0, spanLength: 8)
        createGap(3)
        createWallSpanWithHeight(1, spanLength: 50)
        createGap(20)
    }
    
    func createGap(spanLength: NSInteger) {
        for i in 0..spanLength {
            _walls.addSlice(SliceType.Gap)
        }
    }
    
    func createWallSpanWithHeight(
            heightIndex: Int,
        var spanLength:  Int,
            noFront:     Bool = false,
            noBack:      Bool = false)
    {
        if !noFront && spanLength > 0 {
            addWallFront(heightIndex)
            spanLength--
        }
        
        let midSpanLength: Int = spanLength - (noBack ? 0 : 1)
        if midSpanLength > 0 {
            addWallMid(heightIndex, spanLength: midSpanLength)
            spanLength -= midSpanLength
        }
        
        if !noBack && spanLength > 0 {
            addWallBack(heightIndex)
        }
    }
    
    func createSteppedWallSpanWithHeight(var heightIndex: Int, spanALength: Int, spanBLength: Int) {
        if heightIndex < 2 {
            heightIndex = 2
        }
        
        createWallSpanWithHeight(heightIndex, spanLength: spanALength, noFront: false, noBack: true)
        addWallStep(heightIndex - 2)
        createWallSpanWithHeight(heightIndex - 2, spanLength: spanBLength - 1, noFront: true, noBack: false)
    }
    
    func addWallFront(heightIndex: Int) {
        let y = WALL_HEIGHTS[heightIndex]
        _walls.addSlice(SliceType.Front, y: y)
    }
    
    func addWallBack(heightIndex: Int) {
        let y = WALL_HEIGHTS[heightIndex]
        _walls.addSlice(SliceType.Back, y: y)
    }
    
    func addWallMid(heightIndex: Int, spanLength: Int) {
        let y = WALL_HEIGHTS[heightIndex]
        for i in 0..spanLength {
            if i % 2 == 0 {
                _walls.addSlice(SliceType.Window, y: y)
            } else {
                _walls.addSlice(SliceType.Decoration, y: y)
            }
        }
    }
    
    func addWallStep(heightIndex: Int) {
        let y = WALL_HEIGHTS[heightIndex]
        _walls.addSlice(SliceType.Step, y: y)
    }
}